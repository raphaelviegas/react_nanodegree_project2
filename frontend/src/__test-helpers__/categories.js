export const initialState = { categories: [] }

export const categories = [
    {
        name: 'react',
        path: 'react'
    },
    {
        name: 'redux',
        path: 'redux'
    },
    {
        name: 'udacity',
        path: 'udacity'
    }
]