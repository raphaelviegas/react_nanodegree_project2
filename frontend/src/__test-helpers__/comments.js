const uuidv4 = require('uuid/v4')

export const initialState = { comments: [] }

export const comments = [
    {
        id: '894tuq4ut84ut8v4t8wun89g',
        parentId: "8xf0y6ziyjabvozdd253nd",
        timestamp: 1468166872634,
        body: 'Hi there! I am a COMMENT.',
        author: 'thingtwo',
        voteScore: 6,
        deleted: false,
        parentDeleted: false
    },
    {
        id: '8tu4bsun805n8un48ve89',
        parentId: "8xf0y6ziyjabvozdd253nd",
        timestamp: 1469479767190,
        body: 'Comments. Are. Cool.',
        author: 'thingone',
        voteScore: -5,
        deleted: false,
        parentDeleted: false
    }
]

export const getCommentsId = '894tuq4ut84ut8v4t8wun89g'

export const removeComment = '894tuq4ut84ut8v4t8wun89g'

export const voteComment = {
    id: '894tuq4ut84ut8v4t8wun89g',
    voteScore: 10
}

let id = uuidv4()
export const addCommentInfo = {
    parentId: '8xf0y6ziyjabvozdd253nd',
    author: 'Test Author',
    body: 'Test Body',
    id
}

export const editCommentInfo = {
    id: '894tuq4ut84ut8v4t8wun89g',
    editContent: {
        body: 'Edited comment body',
        timestamp: Date.now()
    }
}