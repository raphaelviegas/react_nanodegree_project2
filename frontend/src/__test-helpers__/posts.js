const uuidv4 = require('uuid/v4')

export const initialState = {
    posts: [],
    sortByOption: '-voteScore',
    invalidPost: false,
    loadingPosts: true,
}

export const posts = [
    {
        id: '8xf0y6ziyjabvozdd253nd',
        timestamp: 1467166872634,
        title: 'Udacity is the best place to learn React',
        body: 'Everyone says so after all.',
        author: 'thingtwo',
        category: 'react',
        voteScore: 10,
        deleted: false,
        commentCount: 2
    },
    {
        id: '6ni6ok3ym7mf1p33lnez',
        timestamp: 1468479767190,
        title: 'Learn Redux in 10 minutes!',
        body: 'Just kidding. It takes more than 10 minutes to learn technology.',
        author: 'thingone',
        category: 'redux',
        voteScore: -5,
        deleted: false,
        commentCount: 0
    }
]

export const removePost = '8xf0y6ziyjabvozdd253nd'

export const votePost = {
    id: '8xf0y6ziyjabvozdd253nd',
    voteScore: 10
}

let id = uuidv4()
export const addPostInfo = {
    author: "Test Author",
    body: "Test Body",
    category: "Test",
    title: "Test Title",
    timestamp: Date.now(),
    id
}

export const editPostInfo = {
    id: '8xf0y6ziyjabvozdd253nd',
    editContent: {
        title: 'Edited post title',
        body: 'Edited post body'
    }
}