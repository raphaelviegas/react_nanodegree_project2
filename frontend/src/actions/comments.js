import * as ServerAPI from '../helpers/ServerAPI'

export const LOAD_COMMENTS = 'LOAD_COMMENTS'
export const REMOVE_COMMENT = 'REMOVE_COMMENT'
export const VOTE_COMMENT = 'VOTE_COMMENT'
export const ADD_COMMENT = 'ADD_COMMENT'
export const EDIT_COMMENT = 'EDIT_COMMENT'

//Local state
export function loadComments (comments) {
  return {
    type: LOAD_COMMENTS,
    payload: comments
  }
}
export function voteComment (id, voteScore) {
  return {
    type: VOTE_COMMENT,
    payload: {id, voteScore }
  }
}

export function removeComment (id) {
  return {
    type: REMOVE_COMMENT,
    payload: id
  }
}

export function addComment(comment) {
  return {
    type: ADD_COMMENT,
    payload: comment
  }
}

export function editComment (id, editContent) {
  return {
    type: EDIT_COMMENT,
    payload: {
      id,
      editContent
    }
  }
}

//Fetch requests
export const fetchComments = (postId) => dispatch => (
  ServerAPI.getComments(postId)
    .then(comments => dispatch(loadComments(comments)))
)

export const fetchVoteComment = (id, option) => dispatch => (
  ServerAPI.vote('comments', id, option)
    .then(comment => dispatch(voteComment(comment.id, comment.voteScore)))
)

export const fetchRemoveComment = (id) => dispatch => (
  ServerAPI.remove('comments', id)
    .then(comment => dispatch(removeComment(comment.id)))
)

export const fetchAddComment = (commentInfo) => dispatch => (
  ServerAPI.add('comments', commentInfo)
    .then(comment => dispatch(addComment(comment)))
)

export const fetchEditComment = (id, editContent) => dispatch => (
  ServerAPI.edit('comments', id, editContent)
    .then(()=> dispatch(editComment(id, editContent)))
)
