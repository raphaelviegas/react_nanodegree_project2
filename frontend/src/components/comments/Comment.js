import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Moment from 'react-moment';
import VoteScore from '../shared/VoteScore'

class Comment extends Component {
    static propTypes = {
        commentInfo: PropTypes.object.isRequired,
        onRemoveComment: PropTypes.func.isRequired,
        onEditComment: PropTypes.func.isRequired,
    }

    removeComment(id) {
        this.props.onRemoveComment(id)
    }

    editComment(id, body) {
        this.props.onEditComment(id, body)
    }

    render() {
        const { author, timestamp, body, voteScore, id } = this.props.commentInfo
        return (
            <li>
                <div className="card">
                    <div className="card-content">
                        <span className="card-title comment-title">Author: {author}</span>
                        <span className="card-date comment-date"><Moment format="DD/MMM/YY">{timestamp}</Moment></span>
                        <p className="card-body comment-body">{body}</p>
                    </div>
                    <div className="card-action comment-action">
                        <span className="comment-action-edit" onClick={() => this.editComment(id, body)}>Edit</span>
                        <span className="comment-action-remove" onClick={() => this.removeComment(id)}>Delete</span>
                    </div>
                    <div className="comment-vote-score">
                        <VoteScore size={12} score={voteScore} type='comment' id={id} />
                    </div>
                </div>
            </li>
        )
    }
}

export default Comment
