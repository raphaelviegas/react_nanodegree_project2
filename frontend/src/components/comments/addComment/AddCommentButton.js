import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddCommentModal from './AddCommentModal'

export class AddCommentButton extends Component {
    static propTypes = {
        parentId: PropTypes.string.isRequired,
    }

    state = {
        modalOpen: ''
    }

    openModal = () => this.setState(() => ({ modalOpen: 'modal-open' }))
    closeModal = () => this.setState(() => ({ modalOpen: '' }))

    render() {
        const { modalOpen } = this.state
        const { parentId } = this.props
        return (
            <div>
                <button
                    className="btn-flat blue-grey lighten-5 add-comment-btn"
                    onClick={() => this.openModal()}
                >Add nem Comment</button>
                <AddCommentModal
                    modalOpen={modalOpen}
                    onCloseModal={this.closeModal}
                    parentId={parentId}
                />
            </div>
        )
    }
}

export default AddCommentButton
