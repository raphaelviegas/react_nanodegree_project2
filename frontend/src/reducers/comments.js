import {
    LOAD_COMMENTS, 
    REMOVE_COMMENT, 
    VOTE_COMMENT,
    EDIT_COMMENT,
    ADD_COMMENT
} from '../actions/comments'

function commentsReducer (state = {
    comments: [],
}, action) {
    switch (action.type) {
        case LOAD_COMMENTS:
            state = {
                ...state,
                comments: action.payload
            }
            return state
        case REMOVE_COMMENT:
            state ={
                ...state,
                comments: state.comments.filter(comment => comment.id !== action.payload)
            }
            return state
        case VOTE_COMMENT:
            state ={
                ...state,
                comments: state.comments.map(comment => {
                    if (comment.id === action.payload.id) {
                        comment.voteScore = action.payload.voteScore
                    }
                    return comment
                })
            }
            return state
        case ADD_COMMENT:
            let comments = state.comments.concat([action.payload])
            state = {
                ...state,
                comments
            }
            return state
        case EDIT_COMMENT:
            state = {
                ...state,
                comments: state.comments.map(comment => {
                    if (comment.id === action.payload.id) {
                        comment.timestamp = action.payload.editContent.timestamp
                        comment.body = action.payload.editContent.body
                    }
                    return comment
                })
            }
            return state
        default:
            return state
    }
}

export default commentsReducer