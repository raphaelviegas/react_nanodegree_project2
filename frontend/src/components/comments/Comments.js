import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { fetchComments, fetchRemoveComment } from '../../actions/comments'
import MaterialIcon from 'material-icons-react'
import sortBy from 'sort-by'
import AddCommentButton from './addComment/AddCommentButton'
import EditCommentModal from './editComment/EditCommentModal'
import Comment from './Comment'
import M from "materialize-css/dist/js/materialize.min.js"

export class Comments extends Component {
    static propTypes = {
        parentId: PropTypes.string.isRequired,
        comments: PropTypes.array.isRequired
    }

    state = {
        sortByOption: '-voteScore',
        commentId: '',
        commentBody: '',
        editModalOpen: ''
    }

    showEditModal = (id, body) => this.setState(
        () => ({ editModalOpen: 'modal-open', commentId: id, commentBody: body })
    )

    hideEditModal = () => this.setState(
        () => ({ editModalOpen: '', commentId: '', commentBody: '', })
    )

    componentDidMount() {
        this.props.fetchComments(this.props.parentId)
        let elems = document.querySelectorAll('.dropdown-trigger');
        // eslint-disable-next-line
        let instances = M.Dropdown.init(elems, { alignment: 'right', constrainWidth: false, hover: false });
    }

    removeComment = (id) => {
        this.props.fetchRemoveComment(id)
    }

    changeSortCriteria = (option) => {
        switch (option) {
            case 'voteScore':
                this.setState({ sortByOption: 'voteScore' })
                break
            case '-voteScore':
                this.setState({ sortByOption: '-voteScore' })
                break
            case 'timestamp':
                this.setState({ sortByOption: 'timestamp' })
                break
            case '-timestamp':
                this.setState({ sortByOption: '-timestamp' })
                break
            default:
                break
        }
    }

    render() {
        const { comments, parentId } = this.props
        const commentCount = comments.length
        const { sortByOption, commentId, commentBody, editModalOpen } = this.state
        //Sort By
        comments.sort(sortBy(sortByOption))
        return (
            <div className="post-comments">
                <div className="post-comments-title">
                    Comments: {commentCount}
                    <div className="order-by">
                        <a className='dropdown-trigger' data-target='dropdown1'><MaterialIcon icon="sort" size={22} /></a>
                        <ul id='dropdown1' className='dropdown-content'>
                            <li><span>Vote Score</span></li>
                            <li><a className="order-button" onClick={() => this.changeSortCriteria('voteScore')}>Ascending</a></li>
                            <li><a className="order-button" onClick={() => this.changeSortCriteria('-voteScore')}>Descending</a></li>
                            <li className="divider" tabIndex="-1"></li>
                            <li><span>Date</span></li>
                            <li><a className="order-button" onClick={() => this.changeSortCriteria('timestamp')}>Ascending</a></li>
                            <li><a className="order-button" onClick={() => this.changeSortCriteria('-timestamp')}>Descending</a></li>
                        </ul>
                    </div>
                </div>
                <div className="divider"></div>
                {commentCount === 0
                    ? <div className="noComments"><h6>No comments to show</h6></div>
                    : <ul className="comments">
                        {comments.map(comment => (
                            <Comment
                                commentInfo={comment}
                                key={comment.id}
                                onRemoveComment={this.removeComment}
                                onEditComment={this.showEditModal}
                            />
                        ))}
                    </ul>
                }
                <AddCommentButton parentId={parentId} />
                <EditCommentModal
                    modalOpen={editModalOpen}
                    commentId={commentId}
                    commentBody={commentBody}
                    onCloseModal={this.hideEditModal}
                />
            </div>
        )
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        parentId: ownProps.parentId,
        comments: state.commentsReducer.comments
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchComments: (parentId) => dispatch(fetchComments(parentId)),
        fetchRemoveComment: (id) => dispatch(fetchRemoveComment(id))
    }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Comments));