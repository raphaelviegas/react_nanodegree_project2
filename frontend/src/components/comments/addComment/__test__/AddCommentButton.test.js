import React from 'react'
import { shallow } from 'enzyme'
import AddCommentButton from '../AddCommentButton'

import { addCommentInfo } from '../../../../__test-helpers__/comments'

describe('<AddCommentButton />', () => {
    const setup = {
        parentId: addCommentInfo.parentId,
    }
    let wrapper
    beforeEach(() => {
        wrapper = shallow(<AddCommentButton {...setup} />)
    })

    it('Should render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('Should render an empty initial state', () => {
        expect(wrapper.state()).toEqual({ modalOpen: '' })
    })

    it('Should render an AddCommentModal component', () => {
        expect(wrapper.find('AddCommentModal').length).toBe(1)
    })

    it('Should update state when button is clicked', () => {
        expect(wrapper.state()).toEqual({ modalOpen: '' })
        wrapper.find('button').simulate('click')
        expect(wrapper.state()).toEqual({ modalOpen: 'modal-open' })
    })

    it('Should update state when method openModal is called', () => {
        expect(wrapper.state()).toEqual({ modalOpen: '' })
        wrapper.instance().openModal()
        expect(wrapper.state()).toEqual({ modalOpen: 'modal-open' })
    })

    it('Should update state when method closeModal is called', () => {
        expect(wrapper.state()).toEqual({ modalOpen: '' })
        wrapper.instance().openModal()
        expect(wrapper.state()).toEqual({ modalOpen: 'modal-open' })
        wrapper.instance().closeModal()
        expect(wrapper.state()).toEqual({ modalOpen: '' })
    })

    it('Should call method openModal when .add-comment-btn is clicked', () => {
        const mockopenModal = jest.fn()
        wrapper.instance().openModal = mockopenModal
        expect(mockopenModal).toHaveBeenCalledTimes(0)
        wrapper.find('.add-comment-btn').simulate('click')
        expect(mockopenModal).toHaveBeenCalledTimes(1)
    })
})