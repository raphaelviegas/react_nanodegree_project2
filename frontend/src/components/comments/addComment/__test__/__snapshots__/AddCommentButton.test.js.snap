// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCommentButton /> Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddCommentButton
    parentId="8xf0y6ziyjabvozdd253nd"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="btn-flat blue-grey lighten-5 add-comment-btn"
          onClick={[Function]}
>
          Add nem Comment
</button>,
        <AddCommentModal
          modalOpen=""
          onCloseModal={[Function]}
          parentId="8xf0y6ziyjabvozdd253nd"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add nem Comment",
          "className": "btn-flat blue-grey lighten-5 add-comment-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add nem Comment",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "modalOpen": "",
          "onCloseModal": [Function],
          "parentId": "8xf0y6ziyjabvozdd253nd",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="btn-flat blue-grey lighten-5 add-comment-btn"
            onClick={[Function]}
>
            Add nem Comment
</button>,
          <AddCommentModal
            modalOpen=""
            onCloseModal={[Function]}
            parentId="8xf0y6ziyjabvozdd253nd"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add nem Comment",
            "className": "btn-flat blue-grey lighten-5 add-comment-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add nem Comment",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "modalOpen": "",
            "onCloseModal": [Function],
            "parentId": "8xf0y6ziyjabvozdd253nd",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
